### PR TITLE
Support multipule proxies

### PR DESCRIPTION
**Pull request by stewartmatheson**
Support more than just rubygems.org as a proxy location

More detail in https://github.com/geminabox/geminabox/issues/198

### CODE REVIEW

**Comment by natebird** [comment]
:+1: I was just about to start working on this exact feature. Thanks!

**Comment by stewartmatheson** [comment]
Hey no worries @natebird.  We have been using for around a week now and it's been quite smooth. Let me know if you run in to any unexpected issues.

**Comment by bhechinger** [comment]
I will set it up now and let you know how it goes.

**Comment by bhechinger** [comment]
Ok, small snag, how do I add a source that is https and has user/pass basic auth?

putting:
  - https://user:pass@site.com/

results in:

`  Message from application: undefined method `rubygems_proxy=' for Geminabox:Module (NoMethodError)`

**Comment by bhechinger** [comment]
Well, I'm not sure what I did, but putting that back in it doesn't explode anymore.

It doesn't seem to fetch from that external source, however.

File: lib/geminabox/proxy/hostess.rb [hunk 1]
```diff
@@ -49,22 +49,37 @@ def serve
...
+            raise Geminabox::Error.new("Cant find ")
+          end
+
           GemStore.create(IncomingGem.new(StringIO.new(content)))
         end
+      end
 
+      def get_gem_from_external_source(source)
+        #external_source_url = 'http://production.cf.' + source
+        external_source_url = 'http://' + source
```
Comment (bhechinger) [inline lib/geminabox/proxy/hostess.rb]: This is what's tripping me up, isn't it? Can you just have the http:// or https:// set in the config file instead?

**Comment by stewartmatheson** [comment]
@bhechinger. humm agreed. This should not be hard coded. There where some places in the existing GIAB code that used legacy rubygems end points. The rubygems.org people where good enough to set up forwarding so allowing the protocol in the config should not be too hard. I will write a patch on the weekend.

**Comment by bhechinger** [comment]
Fantastic, thanks!

**Comment by stewartmatheson** [comment]
OK I have updated the branch. You should be able to put http or https depending on what protocol you wish to use. I did some basic testing but not much. Take it for a spin and let me know.

**Comment by bhechinger** [comment]
It seems to work perfectly, thanks!!

One small niggle (and this isn't a major issue, just something I noticed) is it doesn't seem to keep copies of the gem not from rubygems.

```
require "rubygems"
require "geminabox"

Geminabox.data = "/apt-mirror/geminabox"
Geminabox.allow_remote_failure = true

run Geminabox::Server
```

so it seems allow_remote_failure isn't getting properly applied to non-rubygems sources?

File: lib/geminabox/external_dependency.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      def for(*gems)
+        source_list = []
+        Sources.external_sources.each do |source|
+          url = [ external_uri(source), '?gems=', gems.map(&:to_s).join(',') ].join
+          body = Geminabox.http_adapter.get_content(url)
+          source_list << JSON.parse(body)
+        end
+        source_list.flatten.uniq
+      rescue Exception => e
+        return [] if Geminabox.allow_remote_failure
```
Comment (stewartmatheson) [inline lib/geminabox/external_dependency.rb]: Looks like the logic is not quite right here. Currently if any of the remotes fail the entire for method will return an empty array. In reality I would like to see fail over for each source provided allow_remote_failure = true.

**Comment by stewartmatheson** [comment]
Hey @bhechinger. Another good catch.  I will update the logic next chance I get.

**Comment by stewartmatheson** [comment]
@bhechinger  hey can you also confirm a bug that seems to prevent the GIAB checking proxy sources for updated gems. It seems that currently once it has any version of a gem it will never touch proxies again looking for a newer version. Seems to be happening here but we have a proxy so I can't tell if its an issue with that or the code it's self.

**Comment by bhechinger** [comment]
Yeah, let me see what I can do. I should be able to upload an older version of one I'm assuming.

**Comment by natebird** [comment]
Hi @stewartmatheson, @bhechinger and I are working together on this. Is there anything that would cause dependencies to fail to be pulled from Rubygems? I can install the gem directly but dependencies fail to install.

```
> gem install mandrill-api
ERROR:  While executing gem ... (Gem::UnsatisfiableDependencyError)
    Unable to resolve dependency: 'mandrill-api (= 1.0.53)' requires 'excon (< 1.0, >= 0.16.0)’
```

**Comment by stewartmatheson** [comment]
@bhechinger Must have been some network issues I was having because my net is really crappy here in Australia. I have updated the pull request that should update the issue with allowing failover.  I am wondering if it's best to ignore allow remote failure in this case as I can't think of a reason to fail on the first remote source with out trying any others. Thoughts?

@natebird  I have been unable to reproduce this although I do remember seeing a similar issue after first publishing this pull request. From memory it was the same error output that you posted above.  I tried again and the issue went away. At that stage I remember suspecting a network issue. If the external source is slow to respond the rubygems client could be timing out too. Can you try and reproduce this error from a different machine/network?

```
➜  giab-test  gem install mandrill-api
Fetching: mandrill-api-1.0.53.gem (100%)
Successfully installed mandrill-api-1.0.53
Parsing documentation for mandrill-api-1.0.53
Installing ri documentation for mandrill-api-1.0.53
Done installing documentation for mandrill-api after 0 seconds
1 gem installed
➜  giab-test  gem sources
*** CURRENT SOURCES ***

http://localhost:9292/
```

**Comment by freibuis** [comment]
:+1:   i would love it to point to our internal mirror of ruby-gems.org.

**Comment by freibuis** [comment]
Could this be used to point a local file system copy of rubygems? instead of using http?

**Comment by stewartmatheson** [comment]
@freibuis  Hi. Not quite sure I understand your use case. Why not just setup GIAB internally? With this branch you can proxy to where ever you need it (rubygems.org rails-assets.org etc etc)? The first time you pull gems GIAB will cache them so they will live internally.  I'm not sure what you get by having an extra internal mirror of ruby gems in this use case.

GIAB maintains it's database in the file system. It's written to a data folder along with metadata of the current gems it has in it's database. This database is not really intended for outside use so I would not recommend accessing this data folder directly. Again I'm not quite sure what the value of accessing gems from the file system is compared to accessing them via HTTP.

**Comment by freibuis** [comment]
@stewartmatheson  We do run GIAB internally but we would like to point to our local cached copy of rubygems.org. we have been adding our local GIAB + our local cache of rubygems.org into our Gemfile but this causes problems. we then started to use GIAB to proxy so we only need the GIAB source url.. but this is a problem..  if ruby gems goes down or our internet access goes down, bundle goes down with it...   basically we would like GIAB to point to our cache, from what we see from this pull request we could replace rubygems.org infavour of our localcache.

The reason for wanting to access the cache via filesystem is that we have the cache on the same box as GIAB..  instead of doing a http pull for the gem (which is slow) it could point directly via filesystem to the gem file thus speeding the request. side effect would be that HTTP timeouts would not occur which happens a lot with GIAB

**Comment by stewartmatheson** [comment]
HTTP timeouts will occur due to network conditions. This has nothing to do with the protocol it's self or the GIAB implementation. If you do a HTTP call to the same box you should not get any timeouts.

I also doubt you would see any performance increase pulling gems via the file system even if it where possible. When running bundle on a non trivial gem set under reasonable network conditions rubygems and bundler will spend the vast majority of its time unpacking gems and compiling native binaries. Very little time is actually spent reaching out to the network and pulling gems down. 

I'm not sure how your pulling down the whole rubygems.org database however you might be able to trick GIAB in to caching rubygems.org for you. If you have a list of all gems on rubygems.org call "/gems/:gem_name" on your GIAB instance for each entry in that list. When GIAB gets this incoming request it will first check it's own cache. If it can't find the gem it will fail over and check each proxy source. Once it finds the gem in the proxy source it will cache it locally. You can set your script up to run over night each day. I would consider using threads for this job though might take a while to go though one by one.

Because this branch supports multiple proxies you can put rails-assets in your gem list too. The real advantage to this is that you now only ever have one gem source to be worried about. This was the whole idea for this branch as I am behind a big corporate proxy.

**Comment by Napolskih** [comment]
Hi
I am very interested in this functionality. Thank you in advance. :green_heart: 
What is the status of this work?

**Comment by stewartmatheson** [comment]
It's working. We have been running this at my work now ever since I opened
the pull request and have not had any issues. Take it for a spin and let me
know if you run in to any issues.

**Comment by Napolskih** [comment]
Excellent!
I will try to set up a server on my work with our forks.
Please clarify the question. Do I understand correctly that while the proxying server running in proxy mode and in normal mode? Is it possible to simultaneously store and give their gems and be a proxy?

**Comment by stewartmatheson** [comment]
Yes. You can still publish gems to gem in a box. These gems will be made
available to you in the same way that any gem is available on any proxy
source you add. When clients connect this branch will first look with in
the internal gem set and then try and resolve any missing gems externally
with the proxies. This is all done in the background so your clients never
know weather the gem was published locally or fetched via a proxy. It's
worth noting that once a gem is fetched via a proxy it's cached locally so
the next client does not have to wait for the gem to be fetched again.

**Comment by Napolskih** [comment]
We tried to update our geminabox, your version with a fork
https://github.com/stewartmatheson/geminabox/tree/support-multipule-proxies

Immediately we got an error
http://pastebin.com/2E3PdSFf
http://pastebin.com/sTNiezkN

gem, which falls, was released today and uploaded on the new version.

**Comment by stewartmatheson** [comment]
You seem to have more than one proxy source configured in your client. This
should not be needed. Ideally you only have one remote configured on your
client and let the branch do all the proxying for you.

The buffer error is odd. I have never seen that before. Do any other gems
work from that remote source? It's possible the gem sever is doing
something that GIAB does not expect. Can you test pulling a different gem
from that same remote?

**Comment by Napolskih** [comment]
Everything works, but not stable.
It is sometimes said that the gem is not found. restart several times. Version Bandler and then work to change. Big gemset - 320 gems.
Try another. Look. About one source realized.
It can not be linked to a problem with multithreading bundler option -j 4 and simultaneous re-indexing, for example?

Now the errors that are in the log:
http://pastebin.com/fFFZ9WGd

**Comment by stewartmatheson** [comment]
Where have you set up your data directory? Can you also post your list of
proxies?

**Comment by Napolskih** [comment]
```
external_sources:
    - https://gems.railsc.ru
    - https://rubygems.org
    - https://rails-assets.org
```

https://gems.railsc.ru - internal private stable geminabox

Geminabox.data = "/home/gems/gemsproxy/data" - another own data dir (not gems.railsc.ru)

indexed, such as running.
periodically there are messages gem
```
Could not find rack-cache-1.2 in any of the sources
Could not find rake-10.1.1 in any of the sources
Could not find gem 'truncate_html (~> 0.9.2) ruby' in the gems available on this machine.
```

It helps to restart and launch another version bundler
1.10.6
1.7.6
1.7.15
Everything is complicated by the fact that we have a large gemset and bundler have errors and it can properly resolve dependencies. With rubygems.railsc.ru, which is on the old masters, like works better.

**Comment by Napolskih** [comment]
new errors
http://pastebin.com/BdHNVxKP

**Comment by stewartmatheson** [comment]
Can you paste you gemfile as well as all versions of ruby your running
including the server for GIAB

**Comment by stewartmatheson** [comment]
One other thing I just thought of is system resources. They could be the cause of the buffer errors. Missing tmp filed could be also caused by the OS trying to free up space. How much disk space do you have free on your server? How much free memory do you have? What server front end are you running? Nginx? Apache? What server OS?

**Comment by Napolskih** [comment]
```
Linux xxxx 3.13.0-61-generic #100-Ubuntu SMP Wed Jul 29 11:21:34 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`
front - nginx
free space 9 gb
ram - ok
monitoring in zabbix - ok
oom killer - ok
```

On the same machine, a lot of productive resources, including the main geminabox.
Look, a test further.
Heme file Of course, I can send you. But there are gems half private. I send to e-mail.

**Comment by stewartmatheson** [comment]
How are you serving up gems from that box? Is it using GIAB as well? What version?

**Comment by Napolskih** [comment]
Main storage - GIAB 0.12.4

**Comment by stewartmatheson** [comment]
Looks like you have done everything right. I can't see any issues with your setup. The buffer errors your getting seem to be the source of the issue I would look at trying to solve these first. I would try and understand how much io is happening on the server. Is there too may writes or reads happening when your pulling gems. I would then look at running the two different gem servers on two different boxes. Another thing I would look at is zlib. Has that been compiled correctly? Is it at the latest version? Can you update? Sorry I can't be more exact but at this stage I think your issues are deeper than the ruby code.

**Comment by cwninja** [comment]
I'm a bit concerned about the direction this takes us, as hiding the multiple origins from the consumer prevents us from disambiguating gems of the same name from different sources.

For example, if I add rails-assets-bootstrap to rubygems.org, you are screwed.

Thoughts?

**Comment by stewartmatheson** [comment]
Yeah. I guess so. The way we are using GIAB it's unlikely this would ever happen so the risk is minimal. You could include an @ in the gem name but yeah... that's kinda silly and I would not even bother coding a feature like that.

The bigger picture here is that the future of rails assets is shaky at best last time I checked (a while ago). There wont be much of a need for this in the future.  I'm more than happy to maintain a fork for anyone who finds this thread and needs multiproxy. Just reply here and I can pull down stream changes in.

**Pull request closed without merging**